### PR TITLE
Add StableLM support

### LINKED ISSUE #549: stabilityai/stablelm-2-1_6b Model
### Model description

Stable LM 2 1.6B is a 1.6 billion parameter decoder-only language model pre-trained on 2 trillion tokens of diverse multilingual and code datasets for two epochs.

#### Model link:
[stablelm-2-1_6b](https://huggingface.co/stabilityai/stablelm-2-1_6b)

### Prerequisites

- [X] The model is supported in Transformers (i.e., listed [here](https://huggingface.co/docs/transformers/index#supported-models-and-frameworks))
- [X] The model can be exported to ONNX with Optimum (i.e., listed [here](https://huggingface.co/docs/optimum/main/en/exporters/onnx/overview))

### Additional information

```python
from transformers import AutoModelForCausalLM, AutoTokenizer
tokenizer = AutoTokenizer.from_pretrained("stabilityai/stablelm-2-1_6b", trust_remote_code=True)
model = AutoModelForCausalLM.from_pretrained(
  "stabilityai/stablelm-2-1_6b",
  trust_remote_code=True,
  torch_dtype="auto",
)

model.cuda()
inputs = tokenizer("The weather is always wonderful", return_tensors="pt").to(model.device)
tokens = model.generate(
  **inputs,
  max_new_tokens=64,
  temperature=0.70,
  top_p=0.95,
  do_sample=True,
)
print(tokenizer.decode(tokens[0], skip_special_tokens=True))
```

### Your contribution

I would love to test this model on the Node.js platform, and if possible, I also want to run it on React Native through the Expo SDK
[react-native-ONNX](https://github.com/hootan09/react-native-ONNX)

### PR DESCRIPTION
**Pull request by D4ve-R**
Adds support for 
- [x] `StableLMForCausalLM`
- [x]  `StableLMForSequenceClassification`
  
Closes #549

### CODE REVIEW

**Comment by D4ve-R** [comment]
StableLM support in Optimum is currently underway.  
https://github.com/huggingface/optimum/pull/1719

**Comment by xenova** [comment]
I know it's still a draft, but I added a commit to get it into a working state :)

Example usage:
```js
import { pipeline } from '@xenova/transformers';

const generator = await pipeline('text-generation', 'Xenova/tiny-random-StableLmForCausalLM')
const output = await generator('hi')
console.log(output);
```

It's just a randomly initialized tiny model, so the output is gibberish. I will export some larger models too.

**Comment by D4ve-R** [comment]
Awesome! 😎

**Comment by xenova** [comment]
Yes that is how you do it - but luckily in this case, we already have a tiny stablelm model: https://huggingface.co/hf-internal-testing/tiny-random-StableLmForCausalLM (check out the [org](https://huggingface.co/hf-internal-testing) for the full list of them)

**Comment by xenova** [comment]
Here are some of the larger models:
- https://huggingface.co/Xenova/stablelm-2-1_6b
- https://huggingface.co/Xenova/stablelm-2-zephyr-1_6b

which we can test with

([see all](https://huggingface.co/models?library=transformers.js&other=stablelm))

**Comment by xenova** [comment]
Seems to work alright! 👍 

```js
import { pipeline } from '@xenova/transformers';

const generator = await pipeline('text-generation', 'Xenova/stablelm-2-zephyr-1_6b')

const prompt = [{'role': 'user', 'content': 'Tell me a joke'}];
const inputs = generator.tokenizer.apply_chat_template(prompt, { add_generation_prompt: true, tokenize: false });

const output = await generator(inputs, { max_new_tokens: 20 })
console.log(output[0].generated_text);
// "<|user|>\nTell me a joke\n<|assistant|>\nWhy did the tomato turn red?\n\nBecause it saw the salad dressing!"
```

**Comment by D4ve-R** [comment]
Really cool! Will turn this into a pr.

**Comment by xenova** [comment]
Thanks! I think the last thing to do is just export and test with some sequence classifier models. Is that something you'd like to work on?

**Comment by xenova** [comment]
Looks like there aren't any stablelm text-classification models on the [HF Hub](https://huggingface.co/models?pipeline_tag=text-classification&other=stablelm&sort=trending) (other than your test of course). So I think it will be a good idea to move that to a separate PR and get this merged?
image

File: src/models.js [hunk 1]
```diff
@@ -5194,6 +5194,62 @@ export class SegformerForSemanticSegmentation extends SegformerPreTrainedModel {
...
+    /**
+     * Calls the model on new inputs.
+     * @param {Object} model_inputs The inputs to the model.
+     * @returns {Promise<SequenceClassifierOutput>} An object containing the model's output logits for sequence classification.
+     */
+    async _call(model_inputs) {
+        return new SequenceClassifierOutputWithPast(await super._call(model_inputs));
+    }
+}
+
```
Comment (xenova) [inline src/models.js:5251]:
```suggestion
```

File: src/models.js [hunk 2]
```diff
@@ -5380,6 +5436,7 @@ const MODEL_FOR_SEQUENCE_CLASSIFICATION_MAPPING_NAMES = new Map([
     ['mbart', ['MBartForSequenceClassification', MBartForSequenceClassification]],
     ['mobilebert', ['MobileBertForSequenceClassification', MobileBertForSequenceClassification]],
     ['squeezebert', ['SqueezeBertForSequenceClassification', SqueezeBertForSequenceClassification]],
+    ['stablelm', ['StableLMForSequenceClassification', StableLMForSequenceClassification]],
```
Comment (xenova) [inline src/models.js:5439]:
```suggestion
```

File: src/models.js [hunk 3]
```diff
@@ -5845,6 +5903,23 @@ export class SequenceClassifierOutput extends ModelOutput {
...
+     * @param {Tensor} output.past_key_values Contains pre-computed hidden-states (key and values in the self-attention blocks)
+     * that can be used (see `past_key_values` input) to speed up sequential decoding.
+     */
+    constructor({ logits, past_key_values }) {
+        super();
+        this.logits = logits;
+        this.past_key_values = past_key_values;
+    }
+}
+
```
Comment (xenova) [inline src/models.js:5922]:
```suggestion
```

**Review by xenova — CHANGES_REQUESTED** [review]
TODO: move stablelm text classification to separate PR

**Comment by D4ve-R** [comment]
Wow, seems like you can read my mind 😆
You're right, will move this to another PR!
My response i had typed in, but not send yet:  


Little update, since there is currently no stablelm model for text classification on the hub, I tried training my own.
Unfortunately I was unable to train on 16GB T4. I will add the code below to train [`stablelm-2-1_6b`](https://huggingface.co/stabilityai/stablelm-2-1_6b) on [`twitter-financial-news-sentiment`](https://huggingface.co/datasets/zeroshot/twitter-financial-news-sentiment), if somebody wants to give it a shot.
I think we should split this in two seperate prs and merged the Seq Classification one when a model is available.

```python
import numpy as np
from datasets import load_dataset
from transformers import AutoTokenizer, DataCollatorWithPadding, AutoModelForSequenceClassification, TrainingArguments, Trainer
import evaluate
from huggingface_hub import login
login()

dataset = "zeroshot/twitter-financial-news-sentiment"
model = "stabilityai/stablelm-2-1_6b"

dataset = load_dataset(dataset)
tokenizer = AutoTokenizer.from_pretrained(model)
tokenizer.pad_token = tokenizer.eos_token
data_collator = DataCollatorWithPadding(tokenizer=tokenizer)
accuracy = evaluate.load("accuracy")

def compute_metrics(eval_pred):
    predictions, labels = eval_pred
    predictions = np.argmax(predictions, axis=1)
    return accuracy.compute(predictions=predictions, references=labels)

def preprocess_function(examples):
    return tokenizer(examples["text"], truncation=True)

tokenized_dataset = dataset.map(preprocess_function, batched=True)
id2label = {0: "Bearish", 1: "Bullish", 2: "Neutral"}
label2id = {"Bearish": 0, "Bullish": 1, "Neutral": 2}

model = AutoModelForSequenceClassification.from_pretrained(
    model,
    num_labels=len(id2label),
    id2label=id2label,
    label2id=label2id,
)
model.config.pad_token_id = model.config.eos_token_id

training_args = TrainingArguments(
    output_dir="stablelm-2-1_6b-sentiment",
    learning_rate=2e-5,
    # TODO: will it fit 16GB?
    per_device_train_batch_size=8,
    per_device_eval_batch_size=8,
    num_train_epochs=2,
    weight_decay=0.01,
    evaluation_strategy="epoch",
    save_strategy="epoch",
    # TODO: will it fit 16GB?
    gradient_accumulation_steps=4,
    gradient_checkpointing=True,
    optim="adamw_bnb_8bit",
    load_best_model_at_end=True,
    push_to_hub=True,
)

trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=tokenized_dataset["train"],
    eval_dataset=tokenized_dataset["validation"],
    tokenizer=tokenizer,
    data_collator=data_collator,
    compute_metrics=compute_metrics,
)

trainer.train()
trainer.push_to_hub()
```

File: src/models.js [hunk 4]
```diff
@@ -5316,6 +5316,47 @@ export class SegformerForSemanticSegmentation extends SegformerPreTrainedModel {
 
 //////////////////////////////////////////////////
 
+//////////////////////////////////////////////////
+// StableLM models
+export class StableLMPreTrainedModel extends PreTrainedModel {
+    /**
+     * Creates a new instance of the `StableLMPreTrainedModel` class.
+     * @param {Object} config The configuration of the model.
```
Comment (xenova) [inline src/models.js:5324]:
```suggestion
// StableLM models
export class StableLmPreTrainedModel extends PreTrainedModel {
    /**
     * Creates a new instance of the `StableLmPreTrainedModel` class.
     * @param {Object} config The configuration of the model.
```

File: src/models.js [hunk 5]
```diff
@@ -5316,6 +5316,47 @@ export class SegformerForSemanticSegmentation extends SegformerPreTrainedModel {
 
 //////////////////////////////////////////////////
 
+//////////////////////////////////////////////////
+// StableLM models
```
Comment (xenova) [inline src/models.js:5320]:
```suggestion
// StableLm models
```

File: src/models.js [hunk 6]
```diff
@@ -5316,6 +5316,47 @@ export class SegformerForSemanticSegmentation extends SegformerPreTrainedModel {
...
+
+/**
+ * The bare StableLM Model transformer outputting raw hidden-states without any specific head on top.
+ */
+export class StableLMModel extends StableLMPreTrainedModel { }
+
+/**
+ * StableLM Model with a `language modeling` head on top for Causal Language Modeling (with past).
+ */
+export class StableLMForCausalLM extends StableLMPreTrainedModel {
```
Comment (xenova) [inline src/models.js:5349]:
```suggestion
/**
 * The bare StableLM Model transformer outputting raw hidden-states without any specific head on top.
 */
export class StableLMModel extends StableLMPreTrainedModel { }

/**
 * StableLm Model with a `language modeling` head on top for Causal Language Modeling (with past).
 */
export class StableLmForCausalLM extends StableLmPreTrainedModel {
```

File: src/models.js [hunk 7]
```diff
@@ -5553,6 +5594,7 @@ const MODEL_WITH_LM_HEAD_MAPPING_NAMES = new Map([
     ['starcoder2', ['Starcoder2ForCausalLM', Starcoder2ForCausalLM]],
     ['falcon', ['FalconForCausalLM', FalconForCausalLM]],
     ['trocr', ['TrOCRForCausalLM', TrOCRForCausalLM]],
+    ['stablelm', ['StableLMForCausalLM', StableLMForCausalLM]],
```
Comment (xenova) [inline src/models.js:5597]:
```suggestion
    ['stablelm', ['StableLmForCausalLM', StableLmForCausalLM]],
```

File: src/models.js [hunk 8]
```diff
@@ -5316,6 +5316,47 @@ export class SegformerForSemanticSegmentation extends SegformerPreTrainedModel {
 
 //////////////////////////////////////////////////
 
+//////////////////////////////////////////////////
+// StableLm models
+export class StableLMPreTrainedModel extends PreTrainedModel {
```
Comment (xenova) [inline src/models.js:5321]:
```suggestion
export class StableLmPreTrainedModel extends PreTrainedModel {
```

File: src/models.js [hunk 9]
```diff
@@ -5316,6 +5316,47 @@ export class SegformerForSemanticSegmentation extends SegformerPreTrainedModel {
...
+    /**
+     * Creates a new instance of the `StableLMPreTrainedModel` class.
```
Comment (xenova) [inline src/models.js:5323]:
```suggestion
     * Creates a new instance of the `StableLmPreTrainedModel` class.
```

File: src/models.js [hunk 10]
```diff
@@ -5316,6 +5316,47 @@ export class SegformerForSemanticSegmentation extends SegformerPreTrainedModel {
...
+        this.num_layers = this.config.num_hidden_layers;
+        this.dim_kv = this.config.hidden_size / this.num_heads;
+    }
+}
+
+/**
+ * The bare StableLM Model transformer outputting raw hidden-states without any specific head on top.
+ */
+export class StableLMModel extends StableLMPreTrainedModel { }
+
```
Comment (xenova) [inline src/models.js:5345]:
```suggestion
/**
 * The bare StableLm Model transformer outputting raw hidden-states without any specific head on top.
 */
export class StableLmModel extends StableLmPreTrainedModel { }

```

**Review by xenova — COMMENTED** [review]
naming nits

File: scripts/convert.py [hunk 11]
```diff
@@ -83,6 +83,10 @@
         'per_channel': False,
         'reduce_range': False,
     },
+    'stablelm': {
+      'per_channel': False,
+        'reduce_range': False,
+    },
```
Comment (xenova) [inline scripts/convert.py:89]:
```suggestion
    'stablelm': {
        'per_channel': False,
        'reduce_range': False,
    },
```

File: src/models.js [hunk 12]
```diff
@@ -5316,6 +5316,47 @@ export class SegformerForSemanticSegmentation extends SegformerPreTrainedModel {
...
+export class StableLmForCausalLM extends StableLmPreTrainedModel {
+    /**
+     * Calls the model on new inputs.
+     * @param {Object} model_inputs The inputs to the model.
+     * @returns {Promise<CausalLMOutput>} An object containing the model's output logits for causal language modeling.
+     */
+    async _call(model_inputs) {
+        return new CausalLMOutputWithPast(await super._call(model_inputs));
+    }
+}
```
Comment (xenova) [inline src/models.js:5358]:
```suggestion
export class StableLmForCausalLM extends StableLmPreTrainedModel { }
```

**Comment by xenova** [comment]
Merged! Thanks for this @D4ve-R! 🤗 


**Example:** Text generation with `Xenova/stablelm-2-zephyr-1_6b`.

```js
import { pipeline } from '@xenova/transformers';

// Create text generation pipeline
const generator = await pipeline('text-generation', 'Xenova/stablelm-2-zephyr-1_6b');

// Define the prompt and list of messages
const prompt = "Tell me a funny joke."
const messages = [
    { "role": "system", "content": "You are a helpful assistant." },
    { "role": "user", "content": prompt },
]

// Apply chat template
const inputs = generator.tokenizer.apply_chat_template(messages, {
    tokenize: false,
    add_generation_prompt: true,
});

// Generate text
const output = await generator(inputs, { max_new_tokens: 20 });
console.log(output[0].generated_text);
// "<|system|>\nYou are a helpful assistant.\n<|user|>\nTell me a funny joke.\n<|assistant|>\nHere's a joke for you:\n\nWhy don't scientists trust atoms?\n\nBecause they make up everything!"
```

**Pull request merged**